datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKey.kt

```diff
@@ -157,37 +210,350 @@ data class CoseKey(
+        @SerialLabel(5)
+        @SerialName("Base IV")
+        @ByteString
+        val baseIv: ByteArray? = null,
+        @SerialLabel(-1)
+        @SerialName("crv")
+        val curve: CoseEllipticCurve? = null,
+        @SerialLabel(-2)
+        @SerialName("x")
+        @ByteString
+        val x: ByteArray? = null,
+        @SerialLabel(-3)
+        @SerialName("y")
+        @ByteString
+        val y: ByteArray? = null,
+        @SerialLabel(-4)
+        @SerialName("d")
+        @ByteString
+        val d: ByteArray? = null
+    ) : SerialContainer {
+        init {
+            if (type != CoseKeyType.EC2) throw IllegalArgumentException("Not an EC key!")
+        }
+
+        override fun toCoseKey() =
+            CoseKey(type, keyId, algorithm, operations, baseIv, CoseKeyParams.EcYByteArrayParams(curve, x, y, d))
+
+    }
+
+    @Serializable
+    private class CoseRsaKeySerialContainer(
+        @SerialLabel(1)
+        @SerialName("kty")
+        val type: CoseKeyType,
+        @SerialLabel(2)
+        @SerialName("kid")
+        @ByteString
+        val keyId: ByteArray? = null,
+        @SerialLabel(3)
+        @SerialName("alg")
+        val algorithm: CoseAlgorithm? = null,
+        @SerialLabel(4)
+        @SerialName("key_ops")
+        val operations: Array<CoseKeyOperation>? = null,
+        @SerialLabel(5)
+        @SerialName("Base IV")
+        @ByteString
+        val baseIv: ByteArray? = null,
+        @SerialLabel(-1)
+        @SerialName("n")
+        val n: ByteArray? = null,
+        @SerialLabel(-2)
+        @SerialName("e")
+        @ByteString
+        val e: ByteArray? = null,
+        @SerialLabel(-4)
+        @SerialName("d")
+        @ByteString
+        val d: ByteArray? = null
+    ) : SerialContainer {
+        init {
+            if (type != CoseKeyType.RSA) throw IllegalArgumentException("Not an RSA key!")
+        }
+
+        override fun toCoseKey() = CoseKey(type, keyId, algorithm, operations, baseIv, CoseKeyParams.RsaParams(n, e, d))
+    }
+
+    override val descriptor: SerialDescriptor
+        get() = CoseKeySerialContainer.serializer().descriptor
+
+    override fun deserialize(decoder: Decoder): CoseKey {
+        val labels = mapOf<String,Long>(
+            "kty" to 1,
+            "kid" to 2,
+            "alg" to 3,
+            "key_ops" to 4,
+            "Base IV" to 5,
+            "n/crv" to -1,
+            "x/e" to -2,
+            "y" to -3,
+            "d" to 4
+        )
+
+        lateinit var type: CoseKeyType
+        var keyId: ByteArray? = null
+        var alg: CoseAlgorithm? = null
+        var keyOps: Array<CoseKeyOperation>? = null
+        var baseIv: ByteArray? = null
+        var n: ByteArray? = null
+        var crv: CoseEllipticCurve? = null
+        var xOrE: ByteArray? = null
+        var y: ByteArray? = null
+        var d: ByteArray? = null
+
+        decoder.decodeStructure(descriptor) {
+            val e=this
```

Comment:
the formatting and indendation of this function seems off and that doesn't help unterstanding it

Reply:
re-check please